src/citrine/resources/data_concepts.py

```diff
@@ -79,56 +85,68 @@ def build(cls, data: dict, session: Session = None):
             An object corresponding to a data concepts resource.
 
         """
+        data_copy = deepcopy(data)
+        # Extract the values that are only for the client.
+        # They will be attached to the deserialized object later.
+        client_only_dict = dict()
+        for client_key in cls._client_keys:
+            val = DataConcepts._pop_field(data_copy, client_key)
+            client_only_dict.__setitem__(client_key, val)
+
         # Running through a taurus loads/dumps cycle validates all of the fields and ensures
         # the object is now a dictionary with a well-understood structure
-        data_copy_dict = loads(dumps(deepcopy(data))).as_dict()
+        # Note that this process modifies data_copy and makes it unsuitable for further use
+        data_copy_dict = loads(dumps(data_copy)).as_dict()
```

Comment:
(non-blocking / not-really-actionable) I know this code was already here (and in numerous other places), but in general json (de)serialization isn't totally nontrivial computationally. Needing to have code like this seems like a sign that we should be leveraging taurus's serialization constructs better.

Reply:
Here's the motivation for this code: in order to construct a Taurus object from a dictionary, the pointed-to objects must be full objects, not more dictionaries. In other words, you can build a process run from
```
{
  'typ': 'process_run',
  'name': 'my process run',
  'spec': <Process Spec object>
}
```
But you can't build it from
```
{
  'typ': 'process_run',
  'name': 'my process run',
  'spec': {
    'typ': 'process_spec',
    'name': 'my process spec'
  }
}
```
Running `loads(dumps(foo)).as_dict` reconstructs a complete object, runs all of Taurus's object creation logic (very important because it does validation _and_ sets up soft links) and then collapses the top-level object into a dictionary _without_ affecting the pointed-to objects. It takes what can be an uncertain input (is it a fully-serialized dictionary or a partially-serialized dictionary or a Taurus object?) and turns it into something well-understood (a Taurus object). The goal was to not burden the client with the details of what Taurus does when it's rehydrating objects, so if anything I'd say we're over-reliant on Taurus's serialization constructs, to the point that we're running into problems if the serialized JSON from the backend _doesn't_ match what Taurus expects.

Reply:
This is not to say that the current method cannot be improved upon, just that there are two things tying our hands:
1. Reconstructing objects in the client must run all of Taurus's object creation code. Examples: https://github.com/CitrineInformatics/taurus/blob/develop/taurus/entity/object/measurement_run.py#L74 (soft links) https://github.com/CitrineInformatics/taurus/blob/develop/taurus/entity/object/has_conditions.py#L20 (list type validation)
2. The client de/ser code was written assuming that the serialized JSON returned by the backend would match the underlying data model

Reply:
My point is that `loads(dumps(...))` is in principle just a very expensive identity function. It may be (and seems like) the case that we are using this function with overloaded argument types, which makes this the most concise way to get the properties you pointed out, but JSON serialization and deserialization, and `deep_copy` are all pretty expensive operations. I think if we are more deliberate about how we use our serialization functions, we can get away with much less work here. We may also need to have different `build` methods if sometimes the input is a dict of primitives and other times it contains complex types.

Reply:
But it's not an identity--it modifies `loads(dumps(foo))` modifies `foo` from a dictionary of dictionaries to a top-level object, and `.as_dict()` then turns it into a dictionary of objects.

Reply:
I understand that it is not literally the identity function in this case. With the typical `json.loads|dumps` though, it would be. The logic we are actually interested in here is what you just said, and not (de)serialization. I think we should have (or use the existing?) more specific functions to do this work, rather than rely on the side effects of these functions, which both alter object structure and convert to/from JSON string.